core/client/src/main/java/com/thousandeyes/sdk/pagination/Paginator.java

```diff
@@ -0,0 +1,131 @@
+package com.thousandeyes.sdk.pagination;
+
+import java.lang.reflect.InvocationTargetException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+import com.thousandeyes.sdk.client.ApiException;
+
+
+
+public class Paginator<T, R> implements Iterable<T> {
+    private static final Pattern CURSOR_PATTERN = Pattern.compile("cursor=([^&]*)");
+    private final Function<String, R> apiCall;
+    private final Function<R, List<T>> dataExtractor;
+
+    public Paginator(Function<String, R> apiCall, Function<R, List<T>> dataExtractor)
+    {
+        this.apiCall = apiCall;
+        this.dataExtractor = dataExtractor;
+    }
+
+    public List<T> getAll() throws ApiException {
+        List<T> allResults = new ArrayList<>();
+        String cursor = null;
+        String nextHref;
+
+        do {
+            R result = apiCall.apply(cursor);
+            allResults.addAll(dataExtractor.apply(result));
+
+            var clazz = result.getClass();
+            try {
+                var getLinks = clazz.getMethod("getLinks", null);
+                var links = getLinks.invoke(result);
+
+                var getNext = links.getClass().getMethod("getNext", null);
+                var next = getNext.invoke(links, null);
+
+                if (next != null) {
+                    var getHref = next.getClass().getMethod("getHref", null);
+                    nextHref = (String) getHref.invoke(next, null);
+                    cursor = extractCursor(nextHref);
+                }
+                else {
+                    cursor = null;
+                }
+            }
+            catch (NoSuchMethodException | InvocationTargetException | IllegalAccessException e) {
+                throw new RuntimeException("Not a paginated response");
+            }
+
+        }
+        while (cursor != null);
+
+        return allResults;
+    }
+
+    private String extractCursor(String href) {
```

Comment:
i'm not sure if `cursor` is the only parameter needed for pagination, i think i recall some apis needing the time window as well (i'm not totally sure though). If this is true we'll need to take all parameters into consideration or just follow the link if it's easier? 🤔

Reply:
yeah, the time range parameters are needed for it to work properly